### PR TITLE
Add karray.h to cxplat

### PR DESCRIPTION
**Pull request by csujedihy**
It's for source consumption and header only atm (i.e. not part of cxplat.lib).

Added basic build and test (naively simple) automation for karray.

### CODE REVIEW

**Review by mtfriesen — APPROVED** [review]

File: inc/karray.h [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+
+    friend class ::triageClass;
+    friend VOID ::initGlobalTriageBlock();
+
+    ULONG m_bufferSize = 0;
+    ULONG m_numElements = 0;
+    T *_p = nullptr;
+};
+
+}
```
Comment (nibanks) [inline inc/karray.h]: nit: please add trailing newline

File: inc/karray.h [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+
+#include "cxplat.h"
+#include <new.h>
+#ifdef KERNEL_MODE
+#include <ntintsafe.h>
+#else
+#include <intsafe.h>
+#define RtlSIZETMult SizeTMult
+#endif
+#include <wil/wistd_type_traits.h>
```
Comment (nibanks) [inline inc/karray.h]: Why does this depend on wil? The MsQuic version doesn't (it doesn't depend on wil at all). I believe MsQuic simply copied the few dependencies locally into this file, which would still be my preference instead of a platform specific dependency.

**Review by nibanks — COMMENTED** [review]
We should at the very least open a bug to automate build/usage of this. I don't like having a file in a repo that isn't even compiled. We will not know if/when it breaks.

**Comment by nigriMSFT** [comment]
Is karray supported on all/most platforms that cxplat supports? What _platform_ abstraction is karray providing? According to the "About" section for this repo, this repo contains C code. ?

To me, it feels like karray doesn't quite fit into the cxplat charter.

**Comment by csujedihy** [comment]
> Is karray supported on all/most platforms that cxplat supports? What _platform_ abstraction is karray providing? According to the "About" section for this repo, this repo contains C code. ?
> 
> To me, it feels like karray doesn't quite fit into the cxplat charter.

If we can remove wil from karray, it would be supported by all platforms I assume.

**Comment by mtfriesen** [comment]
> If we can remove wil from karray, it would be supported by all platforms I assume.

I don't think it will support any OS besides Windows because it includes `intsafe.h` for example. IMO anything that abstracts user platform and kernel platform is sufficiently "cross platform" to be in this repo, though.

File: inc/karray.h (hunk 2 above)
Comment (csujedihy) [reply inc/karray.h]: removed wil dependency

**Comment by csujedihy** [comment]
> We should at the very least open a bug to automate build/usage of this. I don't like having a file in a repo that isn't even compiled. We will not know if/when it breaks.

We now build and test karray for windows.

File: src/test/CxPlatTests.h [hunk 3]
```diff
@@ -92,9 +93,9 @@ LogTestFailure(
 static const GUID CXPLAT_TEST_DEVICE_INSTANCE =
 { 0x3a37b2cb, 0x39a6, 0x426a,{ 0xba, 0xf4, 0x77, 0xd0, 0xed, 0x00, 0x70, 0xb3 } };
 
-#ifndef _KERNEL_MODE
+#ifndef KERNEL_MODE
```
Comment (nibanks) [inline src/test/CxPlatTests.h]: Why did you change this. MsQuic uses `_KERNEL_MODE` because (AFAICT) it's defined by VS/msbuild automatically for kernel mode.

File: src/test/bin/winkernel/cxplattest.kernel.vcxproj [hunk 4]
```diff
@@ -92,12 +92,12 @@
   <ItemDefinitionGroup Condition="'$(Configuration)'=='Debug'">
     <ClCompile>
       <RuntimeLibrary>MultiThreadedDebugDLL</RuntimeLibrary>
-      <PreprocessorDefinitions>CX_PLATFORM_WINKERNEL;SECURITY_KERNEL;SECURITY_WIN32;_DEBUG;DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions>CX_PLATFORM_WINKERNEL;SECURITY_KERNEL;SECURITY_WIN32;_DEBUG;DEBUG;KERNEL_MODE=1;POOL_ZERO_DOWN_LEVEL_SUPPORT=1;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (nibanks) [inline src/test/bin/winkernel/cxplattest.kernel.vcxproj]: You shouldn't need to define `KERNEL_MODE`. `_KERNEL_MODE` should already be defined.

**Comment by nibanks** [comment]
> Is karray supported on all/most platforms that cxplat supports? What _platform_ abstraction is karray providing? According to the "About" section for this repo, this repo contains C code. ?
> 
> To me, it feels like karray doesn't quite fit into the cxplat charter.

Really, `KArray` is a kernel mode replacement for `std::vector`; at least that's how it's used in MsQuic, and I believe will be in XDP. So, the best way to approach this then is likely something like a `CxPlatVector` (C++) abstraction: On user mode platforms, it's trivially typedef'ed for `std::vector`, and for kernel mode, it's based on `KArray`. Then we have a platform agnostic vector to use in our test code (and we can have tests for all platforms as well).

For more context, see `DrillBuffer` from MsQuic:
```c++
#ifdef _KERNEL_MODE
class DrillBuffer : public Rtl::KArray<uint8_t>
{
    public:
    void
    push_back(
        _In_ uint8_t value
        )
    {
        CXPLAT_FRE_ASSERT(append(value));
    }

    const uint8_t* data() const { return &(*this)[0]; }

    size_t size() const { return count(); }

    void
    insert(
        _In_ const iterator &dest,
        _In_ const const_iterator &start,
        _In_ const const_iterator &end
        )
    {
        CXPLAT_FRE_ASSERT(insertAt(dest, start, end));
    }
};
#else
using DrillBuffer = std::vector<uint8_t>;
#endif
```
I'd almost simply recommend copy/paste this and rename `DrillBuffer` to `CxPlatVector`, and rename the file to `CxPlatVector` (but don't hard code to `uint8_t`; keep it templated).

File: src/test/CxPlatTests.h (hunk 3 above)
Comment (csujedihy) [reply src/test/CxPlatTests.h]: Oh, right. I was confused by XDP projects setting KERNEL_MODE=1.

**Review by nibanks — APPROVED** [review]
